libjolie/src/main/java/jolie/lang/parse/module/Finder.java

```diff
@@ -0,0 +1,235 @@
+
+	/**
+	 * @return an array of tokens except last one, which denote the module name
+	 */
+	protected abstract String[] packagesToken();
+
+	/**
+	 * @return an importing module name
+	 */
+	protected String moduleName() {
+		return this.target[ this.target.length - 1 ];
+	}
+
+	/**
+	 * Perform a lookup to a jap filename from basePath
+	 * 
+	 * @param basePath base path for lookup
+	 * @param filename a filename
+	 * @return a new File of jap file, null if file is not found
+	 */
+	protected File japLookup( Path basePath, String filename ) throws FileNotFoundException {
+		Path japPath = basePath.resolve( filename + ".jap" );
+		if( Files.exists( japPath ) ) {
+			return japPath.toFile();
+		}
+		throw new FileNotFoundException( japPath.toAbsolutePath().toString() );
+	}
+
+	/**
+	 * Perform a lookup to a ol filename from basePath
+	 * 
+	 * @param basePath base path for lookup
+	 * @param filename a filename
+	 * @return a new path of ol file, null if file is not found
+	 */
+	protected File olLookup( Path basePath, String filename ) throws FileNotFoundException {
+		Path olPath = basePath.resolve( filename + ".ol" );
+		if( Files.exists( olPath ) ) {
+			return olPath.toFile();
+		}
+		throw new FileNotFoundException( olPath.toString() );
+	}
+
+	/**
+	 * perform lookup to the target module's
+	 * 
+	 * @param basePath base path for lookup
+	 * @return Source object corresponding to the module.
+	 * @throws FileNotFoundException when a module is not found
+	 */
+	protected Source locateModule( Path basePath ) throws FileNotFoundException {
+		Path packagePath = Paths.get( basePath.toString(), this.packagesToken() );
+		File olTargetFile = this.olLookup( packagePath, this.moduleName() );
+		if( !olTargetFile.exists() ) {
+			throw new FileNotFoundException( olTargetFile.toString() );
+		}
+		return new FileSource( olTargetFile );
+	}
+
+}
+
+
+/**
+ * A class represent the finder for relative ImportStatement
+ */
+class RelativePathFinder extends Finder {
+
+	/**
+	 * an URI source of the caller
+	 */
+	private final URI source;
+	private int packagesTokenStartIndex = 0;
+
+	protected RelativePathFinder( String[] target, URI source ) {
+		super( target );
+		assert source != null;
+		this.source = source;
+	}
+
+
+	/**
+	 * resolve path from source, each dot prefix means 1 level higer from the caller path directory
+	 */
+	private Path resolveDotPrefix() {
+		Path sourcePath = Paths.get( source );
+		Path basePath;
+		if( !sourcePath.toFile().isDirectory() ) {
+			basePath = sourcePath.getParent();
+		} else {
+			basePath = sourcePath;
+		}
+		int i = 1;
+		for( ; i < this.target.length - 1; i++ ) {
+			if( target[ i ].isEmpty() ) {
+				basePath = basePath.getParent();
+			} else {
+				break;
+			}
+		}
+		packagesTokenStartIndex = i;
+		return basePath;
+	}
+
+	@Override
+	public Source find() throws ModuleNotFoundException {
+		Path basePath = resolveDotPrefix();
+		try {
+			Source moduleFile = super.locateModule( basePath );
```

Comment:
This code is very hard to follow, because you rely on the side-effect that resolveDotPrefix changes what packagesToken will return inside of super.locateModule.

I'd remove the packagesToken method entirely, and have that resolveDotPrefix returns you a composite object containing both the basePath and the array of "packagesToken" (maybe consider calling it pathParts instead? It took me a while to figure out what packagesToken is).

Then, have `locateModule` take the "packagesToken" thing as a second parameter every time.